include/circt/Support/PrettyPrinter.h

```diff
@@ -0,0 +1,288 @@
+//===- PrettyPrinter.h - Pretty printing ----------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This implements a pretty-printer.
+// "PrettyPrinting", Derek C. Oppen, 1980.
+// https://dx.doi.org/10.1145/357114.357115
+//
+// This was selected as it is linear in number of tokens O(n) and requires
+// memory O(linewidth).
+//
+// See PrettyPrinter.cpp for more information.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef CIRCT_SUPPORT_PRETTYPRINTER_H
+#define CIRCT_SUPPORT_PRETTYPRINTER_H
+
+#include "circt/Support/LLVM.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/SaveAndRestore.h"
+
+#include <cstdint>
+#include <deque>
+#include <vector>
+
+namespace circt {
+namespace pretty {
+
+//===----------------------------------------------------------------------===//
+// Tokens
+//===----------------------------------------------------------------------===//
+
+/// Style of breaking within a group:
+/// - Consistent: all fits or all breaks.
+/// - Inconsistent: best fit, break where needed.
+/// - Never: force no breaking including nested groups.
+enum class Breaks { Consistent, Inconsistent, Never };
+
+/// Style of indent when starting a group:
+/// - Visual: offset is relative to current column.
+/// - Block: offset is relative to current base indentation.
+enum class IndentStyle { Visual, Block };
+
+class Token {
```

Comment:
This is a lot for what this is re:representing and storing these tokens.

Capturing some thoughts, we might eventually want to move to a model where:
1) Only unique `StringToken`s and rare instances of the others are heap-allocated, all the rest can be statically created (`StringToken`'s based on statically-constant string literals, and vast majority of other token instances which are just `ibox2` and such) or otherwise represented cheaply (see `PP` enum).
2) Encode tokens taking advantage of (1) so they're no larger than a pointer in common use and per-token-type memory usage (EndToken just needs `Kind`, etc.) like normal LLVM-y RTTI hierarchy (and drop this tagged union).  The lifetimes of these `Token`s could be moved to a `PPContext` or so (similar to `PPBuilderStringSaver`) but managing their lifetime externally is difficult as-is (imagine `add` taking a token by reference), or at least needs to be decoupled from the current `Listener` callback -- so we don't `clear()` out memory backing tokens /during/ the add referring to one of them.

Just some thoughts to avoid costs incurred from copying them around and managing the `deque`s.